### PR TITLE
[FastISel] Support unreachable with NoTrapAfterNoReturn

### PR DESCRIPTION
**Pull request by nikic**
Currently FastISel triggers a fallback if there is an unreachable terminator and the TrapUnreachable option is enabled (the ISD::TRAP selection does not actually work).

Add handling for NoTrapAfterNoReturn, in which case we don't actually need to emit a trap. The test is just there to make sure there is no FastISel fallback (which is why I'm not testing the case without noreturn). We have other tests that check the actual unreachable codegen variations.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/FastISel.cpp [hunk 1]
```diff
@@ -1851,11 +1851,19 @@ bool FastISel::selectOperator(const User *I, unsigned Opcode) {
     return false;
   }
 
-  case Instruction::Unreachable:
-    if (TM.Options.TrapUnreachable)
+  case Instruction::Unreachable: {
+    if (TM.Options.TrapUnreachable) {
+      if (TM.Options.NoTrapAfterNoreturn) {
+        const auto *Call =
+            dyn_cast_or_null<CallInst>(cast<Instruction>(I)->getPrevNode());
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/FastISel.cpp:1858]: CallBase?

File: llvm/test/CodeGen/X86/no-trap-after-noreturn-fastisel.ll [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+declare void @foo()
+
+define void @noreturn_unreachable() nounwind {
+; CHECK-LABEL: noreturn_unreachable:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    pushq %rax
+; CHECK-NEXT:    callq foo@PLT
+  call void @foo() noreturn
+  unreachable
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/no-trap-after-noreturn-fastisel.ll:13]: Check invoke case?

**Review by aengelke — APPROVED** [review]
LGTM. Was there any concrete use case where this caused a noticeable amount of fallbacks?

**Comment by nikic** [comment]
> LGTM. Was there any concrete use case where this caused a noticeable amount of fallbacks?

Yes, this is one of the more common cause of fallbacks for Rust code. For the test case I'm looking at the distribution looks like this:

```
array(7) {
  ["invoke"]=>
  int(7870)
  ["unreachable"]=>
  int(3160)
  ["landingpad"]=>
  int(630)
  ["args"]=>
  int(566)
  ["ret"]=>
  int(397)
  ["other"]=>
  int(324)
  ["switch"]=>
  int(237)
}
```

**Pull request merged**